File: maximumLengthOfRepeatedSubArray.cpp
/* Given two integer arrays A and B, return the maximum length of an subarray that appears in both arrays. */

class Solution {
public:
    int findLength(vector<int>& A, vector<int>& B) {
        
        int l1 = A.size();
        int l2 = B.size();
        vector<vector<int>> dp(l1+1,vector<int>(l2+1,0));
        int maxm = INT_MIN;
        for(int i=1;i<=l1;i++) {
            
            for(int j=1;j<=l2;j++) {
                if(A[i-1] == B[j-1]) {
                    if(i!=1 && j!=1) {
                        if(A[i-2] == B[j-2]) {
                            dp[i][j] = 1 + dp[i-1][j-1];
                        }
                        else 
                            dp[i][j] = 1;
                    }
                    else 
                        dp[i][j] = 1;
                    
                }
                maxm = max(maxm,dp[i][j]);
            }
        }
        
        return maxm;
    }
};

File: MaximumLengthOfRepeatedSubarray.java
/* Given two integer arrays nums1 and nums2, return the maximum length of a subarray that appears in both arrays. */

class Solution {
    
    int getMaxUtil(int[] nums1, int[] nums2, int idx1, int idx2, int[][] mem) {
        
        if(idx1 >= nums1.length || idx2 >= nums2.length) {
            return 0;
        }
        if(mem[idx1][idx2]!=-1)
            return mem[idx1][idx2];
        int currMax = 0;
        int i = idx1;
        int j = idx2;
        while(i<nums1.length && j<nums2.length && nums1[i] == nums2[j]) {
            i++;
            j++;
            currMax++;
        }
        
        mem[idx1][idx2] = Math.max(currMax,Math.max(getMaxUtil(nums1,nums2,idx1,idx2+1,mem),getMaxUtil(nums1,nums2,idx1+1,idx2,mem)));
        return mem[idx1][idx2];
    }
    public int findLength(int[] nums1, int[] nums2) {
        int[][] mem = new int[nums1.length][nums2.length];
        for(int i=0;i<mem.length;i++) {
            for(int j=0;j<mem[i].length;j++) {
                mem[i][j] = -1;
            }
        }
        return getMaxUtil(nums1,nums2,0,0,mem);
    }
}